### DISCUSSION TITLE
copyExternalImageToTexture and sRGB conversion

### DISCUSSION THREAD

**Discussion by shannon** [gpuweb/gpuweb#4196] [Q&A]
I am having trouble understanding the sRGB conversion and premultiply alpha when using copyExternalImageToTexture. I can't seem to get it to behave the same way as WebGL2 does. (My code falls back to WebGL2 when WebGPU is not supported)

I am using createImageBitmap and supplying the colorSpaceConversion and premultiplyAlpha like so:

```js
createTextureFromImageBitmapData({ data, ...descriptor }) {
        const texture = this.device.createTexture({ ...descriptor, usage: descriptor.usage | TEXTURE_USAGE.COPY_DST });

        createImageBitmap(data, { colorSpaceConversion: 'none', premultiplyAlpha: 'none' }).then(source => {
            this.device.queue.copyExternalImageToTexture({ source }, { texture }, descriptor.size);
        });

        return texture;
}
```
The texture format I am concerned with is rgba8unorm-srgb in this case.

For WebGL2 mode, copyExternalImageToTexture falls back to a simple:

```js
gl.bindTexture(glTarget, glTexture);
gl.texSubImage2D(glTarget, mipLevel, x, y, width, height, webgl2.format, webgl2.type, source);
```

I am NOT using WebGL2 UNPACK_PREMULTIPLY_ALPHA_WEBGL or UNPACK_COLORSPACE_CONVERSION_WEBGL so they are defaults.

The above createImageBitmap settings appear to work perfectly fine for WebGL2. 

I have two samples I am using from the glTF sample models to ensure the encoding is correct

https://github.com/KhronosGroup/glTF-Sample-Models/tree/8e9a5a6/2.0/TextureEncodingTest
https://github.com/KhronosGroup/glTF-Sample-Models/tree/8e9a5a6/2.0/AlphaBlendModeTest

#### Figure 1 (WebGL2 Texture Encoding Test with colorSpaceConversion='none' premultiplyAlpha: 'none')
<img width="625" alt="webgl2_no_conversion_encoding" src="https://github.com/gpuweb/gpuweb/assets/1154610/2d59c4e7-8e89-4a9e-b7b9-c71508c71d7d">

#### Figure 2 (WebGL2 Alpha Blend Test with colorSpaceConversion='none' premultiplyAlpha: 'none')
<img width="625" alt="webgl2_no_conversion_alpha" src="https://github.com/gpuweb/gpuweb/assets/1154610/3887fcee-3ab9-44e4-a6e0-5fdffb27db69">

#### Figure 3 (WebGPU Texture Encoding Test with colorSpaceConversion='none' premultiplyAlpha: 'none')
<img width="625" alt="webgpu_no_conversion_encoding" src="https://github.com/gpuweb/gpuweb/assets/1154610/c5e045a5-7aed-4084-a283-3d68ecb199f9">

#### Figure 4 (WebGPU Alpha Blend Test with colorSpaceConversion='none' premultiplyAlpha: 'none') *Note how the color darkens near the tops of the striped patterns here. This implies that the color has had the alpha premultiplied*
<img width="623" alt="webgpu_no_conversion_alpha" src="https://github.com/gpuweb/gpuweb/assets/1154610/6a02138f-60e3-4ca1-b6ce-7a74decb2d2f">

So I tried different permutations of the createImageBitmap options and copyExternalImageToTexture and nothing seems to work out. If I change the colorSpaceConversion to default, the alpha looks better but the encoding does not.

#### Figure 5 (WebGL2 Texture Encoding with colorSpaceConversion='default' premultiplyAlpha: 'none')
<img width="623" alt="webgl2_default_conversion_encoding" src="https://github.com/gpuweb/gpuweb/assets/1154610/6fe768c5-99f7-43e2-8527-6d4422f2c865">

#### Figure 6 (WebGL2 Alpha Blending Test with colorSpaceConversion='default' premultiplyAlpha: 'none')
<img width="624" alt="webgl2_default_conversion_alpha" src="https://github.com/gpuweb/gpuweb/assets/1154610/17c87389-cfb5-4f9d-82f3-90e35f45f668">

#### Figure 7 (WebGPU Texture Encoding with colorSpaceConversion='default' premultiplyAlpha: 'none')
<img width="625" alt="webgpu_default_conversion_encoding" src="https://github.com/gpuweb/gpuweb/assets/1154610/4ce120cc-603a-41ad-b587-d841a92885dd">

#### Figure 8 (WebGPU Alpha Blend Encoding with colorSpaceConversion='default' premultiplyAlpha: 'none')
<img width="626" alt="webgpu_default_conversion_alpha" src="https://github.com/gpuweb/gpuweb/assets/1154610/e9a04553-f4b6-4071-aa8c-750455c83a50">

### Summary

It seems to me that the first settings should work, and perhaps there is a bug in Chrome, but I am worried that my understanding is incorrect. Should the color be premultiplied in that first `Figure 4` case?

**Comment by shannon** [comment]
Actually I do think it is in fact a bug in Chrome for Windows. It looks fine in Chrome Canary on Android.
#### Figure 9 (WebGPU *on Chrome Canary Android* Alpha Blend Test with colorSpaceConversion='none' premultiplyAlpha: 'none')
![Screenshot_20230623-132625](https://github.com/gpuweb/gpuweb/assets/1154610/c751025c-add3-49a4-9380-faba60bfc262)

#### Figure 10 (WebGPU *on Chrome Canary Android* Texture Encoding Test with colorSpaceConversion='none' premultiplyAlpha: 'none')
![Screenshot_20230623-132810](https://github.com/gpuweb/gpuweb/assets/1154610/c5c48ae9-f598-4ed5-b8ae-a26f177cdcbc)

**Reply by kainino0x** [reply]
We would greatly appreciate a bug report on https://crbug.com/ if you can! Please share it here once you've filed one.

**Reply by shannon** [reply]
I filed the bug here: https://bugs.chromium.org/p/chromium/issues/detail?id=1457649
